### ISSUE TITLE
Allow users to call readRowGroup 

### ISSUE THREAD

**Issue by park-brian** [hyparam/hyparquet#12]
Hello,

First of all, thank you for implementing this package. Would you be able to export the readRowGroup function? This is useful for when we're implementing predicate pushdown to read specific row groups, without needing to calculate rowStart and rowEnd (which sometimes has issues when we need to pass in BigInts greater than MAX_SAFE_INTEGER).

Thank you

**Comment by park-brian** [comment]
I also noticed that you were working on parsing page-level statistics. Does this mean that we might eventually be able to fetch/decode individual data pages without reading the entire RowGroup?

**Comment by platypii** [comment]
Hey @park-brian I just published a new version v0.9.9 which `export`s the `readRowGroup` function. I didn't export it from the package, because right now it feels like a bit more of an internal api. But you should now be able to use it with an import like:

```js
import { readRowGroup } from 'hyparquet/src/read.js'
```

I also just added an optional `rowLimit` argument to the `readRowGroup` which will stop the page parsing early if further rows aren't requested. This is especially nice for rendering the first rows of a file quickly.

Longer term, yes I would like to use all the data available in parquet to make querying for data as fast as possible, including using the ColumnIndex and OffsetIndex. Suggestions and contributions are welcome!

**Comment by park-brian** [comment]
Thank you, that's great! Is it okay to use your parse-indices branch as a reference for parsing the ColumnIndex and OffsetIndex sections? I think this would be all that users need to determine the location of each page (which would let them read the page headers and then call readDataPageV2)

**Comment by platypii** [comment]
Yes please feel free to use it. The intention is to eventually use the indices to do exactly what you are suggesting. If you have any success please let me know!

FYI I just rebased the `parse-indices` branch on top of master.

**Comment by park-brian** [comment]
Thank you, will do! I'll report my findings here.

**Comment by park-brian** [comment]
@platypii I managed to get page-level predicate pushdown working for numeric columns. It's a bit verbose, but it works

https://github.com/park-brian/hyparquet/blob/parse-indices/test/readIndices.test.js

**Comment by park-brian** [comment]
Just a quick note in case others run into this: if we're reading pages that use a dictionary, we need to concatenate the dictionary page + data page before passing the data to readColumn

**Comment by park-brian** [comment]
I can work on adding filters that implement page-level predicate pushdown. It might look a bit like the mongodb query syntax. Eg:

```js
// retrieve a record with an id of 10
{ id: 10 }

// retrieve records with ids in 3, 5, 7, 11
{ id: { $in: [3, 5, 7, 11] } }

// retrieve all records with a value between 20 and 40 (inclusive)
{ value: { $bt: [20, 40] } }

// other operators like $gt, $gte, $lt, $lte, etc can be implemented
{ value: { $gt: 10 } }

// we can even add support for nesting logical operators
{ $and: [ { firstName: 'John' }, { lastName: 'Doe' } ] }
```

Of course, a simpler approach is probably to let the user provide a callback function 

```js
(rowGroupStatistics, pageStatistics) => {/* Some filter logic */}
```

**Comment by platypii** [comment]
That's very cool work @park-brian !

I am interested in adding a `parquetQuery` or similar function. I was starting to think about what the API might look like. I'm not familiar with the mongo syntax but that is super interesting. I'm happy to accept PRs but I do want to be thoughtful about what the API looks like.

**Comment by park-brian** [comment]
Absolutely, good API design saves everyone time and effort. 

It seems that the `readColumnIndex` and `readOffsetIndex` functions work well. Would you accept a PR with tests for these functions? I noticed that the metadata test reads all the test parquet files and compares them with the metadata json files, so I'm assuming this might be the approach you want for testing column/offset indices.

**Comment by park-brian** [comment]
I've found that one way to improve http I/O performance is by specifying multiple ranges in the http range header. Some initial benchmarks show that this decreases response time by 1-2 orders of magnitude (versus making parallel range requests for pages). I guess this should seem obvious in hindsight, but I was not expecting such a massive improvement.

Most web servers are configured to limit headers to around 8k, so it's probably safe to parallelize these multirange requests over 4-6kb chunks.

**Comment by platypii** [comment]
That is a nice find about the performance improvement with multiple ranges.

I already do a trick inside `readRowGroup` where I read the entire RowGroup in one read if it's less than 32mb, otherwise I make one fetch request per-row-group-per-column. I could imagine an aggregation one level higher in `parquetRead` where it determines if it should read multiple consecutive RowGroups all at once. It could do this either as one aggregate single range, or as multiple ranges in one request.

Note that I would want to have multi-ranges as an **option** that can be disabled, because CORS requests only allows a single range for cross-origin requests:

> The Range header is a [CORS-safelisted request header](https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_request_header) when the value is a single byte range. This means that it can be used in cross-origin requests without triggering a preflight request.

**Comment by park-brian** [comment]
Interesting! I guess preflight requests would hurt users with high latency.

There are only a few situations where users would benefit more from multi-range requests rather making requests in parallel (both approaches can be combined, of course). 
- when they're making lots of small range requests (> 100, can happen when fetching individual data pages)
- when they're on a high-latency connection (ironically)

**Comment by platypii** [comment]
> It seems that the `readColumnIndex` and `readOffsetIndex` functions work well. Would you accept a PR with tests for these functions? I noticed that the metadata test reads all the test parquet files and compares them with the metadata json files, so I'm assuming this might be the approach you want for testing column/offset indices.

Yes I would be thrilled to have a PR with the index readers, with tests. I like the idea of testing similarly to the other tests where it reads a snapshot json file with the expected output. Only a small % of the total test files actually contain column and offset indicies, so we probably only need to make new snapshot files for a few of the tests. Do you want to take a look at doing that?

**Comment by park-brian** [comment]
I would be happy to write some tests. Would you be able to rebase parse-indices so it's up to date with master?

**Comment by platypii** [comment]
Rebased parse-indicies

**Comment by park-brian** [comment]
Thanks @platypii !

Testing has revealed an interesting edge case in the metadata.js/convertMetadata function when we're passing in a Uint8Array of 0 length alongside a defined schema (this happened when getting column indices for the adam_genotypes.parquet file).

In this case, is it better to return the original value or undefined? I think that in this context, 0-length arrays are null values.

```js
export function convertMetadata(value, schema) {
  const { type, converted_type, logical_type } = schema
  if (value === undefined || !value.length) return value
  // ...
}
```

or 

```js
export function convertMetadata(value, schema) {
  const { type, converted_type, logical_type } = schema
  if (value === undefined || !value.length) return undefined // parsed as null
  // ...
}
```

Here's what I'm using to read the column indices
```js
const arrayBuffer = await readFileToArrayBuffer(`test/files/${file}`)
const metadata = parquetMetadata(arrayBuffer)

const columnIndices = metadata.row_groups.map((rowGroup) => rowGroup.columns.map((column) => {
  if (column.column_index_offset === undefined || column.column_index_length === undefined) return null
  const columnIndexOffset = Number(column.column_index_offset)
  const columnIndexLength = Number(column.column_index_length)
  const columnIndexArrayBuffer = arrayBuffer.slice(columnIndexOffset, columnIndexOffset + columnIndexLength)
  const columnIndexReader = { view: new DataView(columnIndexArrayBuffer), offset: 0 }
  const schemaPath = getSchemaPath(metadata.schema, column.meta_data?.path_in_schema ?? [])
  return readColumnIndex(columnIndexReader, schemaPath.at(-1)?.element || { name: '' })
}))

```

**Comment by park-brian** [comment]
Resolved with PR #14

**Comment by park-brian** [comment]
@platypii 
By the way, when do you think parse-indices will be merged to master? Just curious if you're planning to do more with this branch.

**Comment by platypii** [comment]
@park-brian I just released version 1.0.0 of hyparquet last night! I wanted to release 1.0.0 before considering the parse-indicies branch, since I was worried it might take some time and effort to stabilize and develop APIs for structured queries.

My plan is to make a V2 branch where we can merge in the parse-indicies branch and continue work on query api. I could release alpha packages to npm with the index parsers if it would be helpful. Open to suggestions though!

Have you thought more about how to implement a query engine?

**Comment by park-brian** [comment]
@platypii Makes sense, it's a good time for a release. Congratulations on the 1.0 release and the HuggingFace grant! 

Let me open a new issue so that we can get the discussion started (we could even try using GitHub Projects). For now, we have a couple of options. What I've been doing locally is passing in a user-defined predicate function which evaluates column/page statistics. However, this is not a great DX so a query API would be much better.

Internally, we would still be applying a predicate to get row groups/column chunks, then pages, and finally rows to retrieve filtered records (as well as fetching adjacent columns the user requests). I like the MongoDB query syntax, since it's concise yet expressive. There are also other query DSLs such as JSONPath or SQL. We could retain the option of letting power users manually specify predicate functions (in fact, this could be implemented first). This would be helpful for users who want complete control over predicate logic (eg: running calculations or querying external data).

There's also the issue of page alignment within columns. Usually, parquet writers will try to make sure that each column chunk in a row group has the same number of pages. If they don't, it makes it a bit harder to fetch adjacent pages in other columns.
